File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        //EDGE CASES PURA CHOD DIYA TUMNE
        //K==1 HEADD==NULL BREAK WALLA CHODA
        //IF IT IS PERFECT MULTIPLE CHODA
        //BEKAR HO
        if(k==1) return head;
        ListNode start=null,end=null,ptr=null,pre=null,fut=null,temp=head,ans=null,p=null;
        ListNode headd=null;
        while(temp!=null){
            start=temp;
            for(int i=0;i<k;i++){
                if(temp!=null)
                  temp=temp.next;
                else{
                 if(headd==null) break;
                 ans.next=start;
                 return headd;
                }
            }
            end=temp;
            ptr=start;
            while(ptr!=end){
                fut=ptr.next;
                ptr.next=pre;
                pre=ptr;
                ptr=fut;
            }
            if(headd==null){
               headd=pre;
               ans=start;
            }
            else{
               ans.next=pre;
                ans=start;
            }
        }
        ans.next=null;
        return headd;
    }
}

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        ans = ListNode(-1)
        ans.next = head

        ptr = head
        c = 0
        while ptr != None:
            c += 1
            ptr = ptr.next
        
        no = c / k

        prev = ans
        cur = head
        c = 0
        
        while cur != None and no != 0:
            start = prev 
            end = cur

            c = 0
            while c <  k and cur != None:
                fut = cur.next
                cur.next = prev
                prev = cur
                cur = fut
                c += 1

            no -= 1
                
            start.next = prev
            end.next = cur
            prev = end
        
        return ans.next